go/private/go_repositories.bzl

```diff
@@ -134,22 +136,51 @@ filegroup(
   name = "go_include",
   srcs = [ "pkg/include" ],
 )
-"""
 
+go_root(
+  name = "go_root",
+  path = "{goroot}",
+)
+"""
 
 def _go_repository_select_impl(ctx):
-  os = ctx.os.name
-  # NOTE: This mapping cannot be table-driven to prevent
-  # Bazel from downloading the other archive.
-  if os == 'linux':
-    goroot = ctx.path(ctx.attr._linux).dirname
-  elif os == 'mac os x':
-    goroot = ctx.path(ctx.attr._darwin).dirname
+  rules_goroot = ctx.os.environ.get("RULES_GOROOT", None)
+
+  # 1. Configure the goroot path
+  if rules_goroot:
+    goroot = ctx.path(rules_goroot)
   else:
-    fail("unsupported operating system: " + os)
+    os_name = ctx.os.name
```

Comment:
long else blocks with more if/else logic can be hard to read.  I think obtain os_name outside the if block (even though in 1 case it goes unused), and then make this all depth 1

os_name = ctx.os.name
if rules_goroot:
  goroot = ctx.path(rules_goroot)
elif os_name == 'linux':
  goroot = ctx.path(ctx.attr._linux).dirname
elif ...

and so forth